### ISSUE TITLE
bug in with primitive

### ISSUE THREAD

**Issue by klag** [jolie/jolie#196]
see test in branch **with_bug**: https://github.com/jolie/jolie/tree/with_bug

**Comment by fmontesi** [comment]
Not a bug: it is valid to write a variable path on multiple lines. So you've basically written .b = .b.c = "yes", which is a valid statement. 

We could change behaviour and forbid people from writing variable paths over multiple lines.. Not sure if this would be better.

**Comment by klag** [comment]
I think it should be considered as a bug. It generates a lot of confusion. It is sufficient to forget the semicolon in a with for getting a bad interpretation of the code

**Comment by fmontesi** [comment]
Just a small clarification: it's incorrect to call it a bug simply because the language works as intended, i.e., variable paths can be expressed over multiple lines, just like in Java and all other languages with the dot operator.

What we have here is that the design of the language plays poorly with the optional semicolon and the fact that assignments are also expressions. So we need to change the design.

Possible solutions:
- forbid to write variable paths over multiple lines in general (this seems weird);
- remove the feature that expressions can contain assignments.

**Comment by klag** [comment]
The problem is related to primitve with because inside its scope, variable paths can start with dot. Thus, it is very common to have more than one expression in different lines which start with dot. But, if we don't put the semicolon, these kind of codes will be always wrong.

**Comment by fmontesi** [comment]
I agree, see my last comment. It's just that solving it in a consistent way is nontrivial, see the possible solutions. 

Another possible solution is to use another symbol than the dot for with-prefixed variable paths (this would break backwards compatibility). Any other ideas?

**Comment by fmontesi** [comment]
Notice that with is typically problematic, like in javascript where they don't recommend using it: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/with

So we might want to make some radical change here. We should identify in which use cases, exactly, with is better than the new inline tree literals (whenever with can be replaced by an inline tree, I'd recommend doing it).

**Comment by fmontesi** [comment]
An example of a non-problematic with syntax:

```
with( my.long.path as a ) {
    a.x = 2
    a.y = 3
}
```

:-)

**Comment by klag** [comment]
This is nice.

**Comment by klag** [comment]
a is an alias?

**Comment by fmontesi** [comment]
Yeah, that's the easy implementation solution.

**Comment by fmontesi** [comment]
@jolie/developers Removing the bug label, per the discussion above. If anybody wants to take a shot at implementing `with( path as x ) { ... }`, please self-assign this.

**Comment by klag** [comment]
Sorry but I think it is a bug not an enhancement.
The ; is mandatory inside the with and optional outside. 
It is a very bad feature. For me it is a bug. I would like to label this as a bug

**Comment by fmontesi** [comment]
Not a bug, technically, but it's so bad that I'm against tagging it as such. It should also be marked as enhancement, since it does not have a reasonable fix without changing the language.

**Comment by fmontesi** [comment]
@jolie/developers, @klag and I had a chat about using .. to disambiguate with.

```jolie
with(x) {
  ..y = 3 // this now assigns x.y = 3
}
```

This would require a slight change in the part of `OLParser`, which currently looks for a `.` prefix and now instead should look for a `..` prefix. `..` should probably be a new scanner token.

If anybody tackles this, please remember to notify this issue.

**Comment by mwallnoefer** [comment]
@Niels-Erik as you have done some parser refactoring, could be an interesting issue?
